kupyna/src/utils.rs

```diff
@@ -44,58 +44,159 @@ fn multiply_gf(x: u8, y: u8) -> u8 {
+            multiply_gf(byte4, MDS_MATRIX[6][4]) ^
+            multiply_gf(byte5, MDS_MATRIX[6][5]) ^
+            multiply_gf(byte6, MDS_MATRIX[6][6]) ^
+            multiply_gf(byte7, MDS_MATRIX[6][7]);
+        output |= (out6 as u64) << 8;
+
+        // Row 7
+        let out7 = multiply_gf(byte0, MDS_MATRIX[7][0]) ^
+            multiply_gf(byte1, MDS_MATRIX[7][1]) ^
+            multiply_gf(byte2, MDS_MATRIX[7][2]) ^
+            multiply_gf(byte3, MDS_MATRIX[7][3]) ^
+            multiply_gf(byte4, MDS_MATRIX[7][4]) ^
+            multiply_gf(byte5, MDS_MATRIX[7][5]) ^
+            multiply_gf(byte6, MDS_MATRIX[7][6]) ^
+            multiply_gf(byte7, MDS_MATRIX[7][7]);
+        output |= out7 as u64;
+
+        state[col] = output;
+    }
 }
 
-pub(crate) fn apply_s_box<const N: usize>(mut state: [[u8; 8]; N]) -> [[u8; 8]; N] {
-    for i in 0..8 {
-        for row in state.iter_mut() {
-            row[i] = SBOXES[i % 4][row[i] as usize];
-        }
+pub(crate) fn apply_s_box<const N: usize>(state: &mut [u64; N]) {
+    for word in state.iter_mut() {
+        // Extract all bytes at once
+        let bytes = word.to_be_bytes();
+
+        // Transform each byte using the appropriate S-box
+        let transformed_bytes = [
+            SBOXES[0][bytes[0] as usize], // row 0 -> S-box 0
+            SBOXES[1][bytes[1] as usize], // row 1 -> S-box 1
+            SBOXES[2][bytes[2] as usize], // row 2 -> S-box 2
+            SBOXES[3][bytes[3] as usize], // row 3 -> S-box 3
+            SBOXES[0][bytes[4] as usize], // row 4 -> S-box 0 (4 % 4 = 0)
+            SBOXES[1][bytes[5] as usize], // row 5 -> S-box 1 (5 % 4 = 1)
+            SBOXES[2][bytes[6] as usize], // row 6 -> S-box 2 (6 % 4 = 2)
+            SBOXES[3][bytes[7] as usize], // row 7 -> S-box 3 (7 % 4 = 3)
+        ];
+
+        // Update the word in-place
+        *word = u64::from_be_bytes(transformed_bytes);
     }
-    state
 }
 
-pub(crate) fn add_constant_xor<const N: usize>(
-    mut state: [[u8; 8]; N],
-    round: usize,
-) -> [[u8; 8]; N] {
-    for (j, row) in state.iter_mut().enumerate() {
+pub(crate) fn add_constant_xor<const N: usize>(state: &mut [u64; N], round: usize) {
+    for (j, word) in state.iter_mut().enumerate() {
         let constant = ((j * 0x10) ^ round) as u8;
-        row[0] ^= constant;
+        let mut bytes = word.to_be_bytes();
+        bytes[0] ^= constant;
+        *word = u64::from_be_bytes(bytes);
     }
-    state
 }
 
-pub(crate) fn add_constant_plus<const N: usize>(
-    mut state: [[u8; 8]; N],
-    round: usize,
-) -> [[u8; 8]; N] {
-    for (j, row) in state.iter_mut().enumerate() {
-        let mut row_as_u64 = u64::from_le_bytes(*row);
+pub(crate) fn add_constant_plus<const N: usize>(state: &mut [u64; N], round: usize) {
+    for (j, word) in state.iter_mut().enumerate() {
+        // Convert to little-endian bytes to match original behavior
+        let mut row_as_u64 = u64::from_le_bytes(word.to_be_bytes());
```

Comment:
You probably should use `from_be` or `from_le` here.

Also, ideally you should not reverse byte order here in the first place and instead just generate a proper constant for LE and BE targets.